### PR TITLE
Add groups/roles support

### PR DESCRIPTION
**Pull request by hugoboos**
First attempt to add groups/roles support.

The Roles are defined in Azkaban and the users in LDAP. The groups connects users to roles.

Groups and roles are defined in an XML file, similar to the `azkaban-users.xml` file.

### CODE REVIEW

**Comment by hugoboos** [comment]
The build fails because of failing unit tests.
I've looked into it. The LDAP server that is being used doesn't work with the group attributes.
The method `addGroupsToUser` that fetch the groups that the user belongs won't work with that LDAP server. There are no attributes `memberof` and `memberuid`.

How should we proceed? Any thoughts?

File: src/main/java/net/researchgate/azkaban/Group.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package net.researchgate.azkaban;
+
+import azkaban.user.Role;
+import azkaban.user.User;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+class Group {
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/Group.java]: You could some Javadoc here describe what this class stands for

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class GroupsLoader {
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: You could some Javadoc here describe what this class stands for

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            }
+
+            groups.put(groupName, group);
+        }
+
+        return groups;
+    }
+
+    private File loadFile(String path) {
+        File file = new File(path);
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: if file is null,  you will get a NPEx should we consider that?

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    private static final String DEFAULT_ROLE_NAME = "admin";
+    private static final String XML_DTD_FILE_PATH = "/azkaban-groups.dtd";
+
+    Map<String, Group> loadFromFile(String filePath) {
+        File file = loadFile(filePath);
+        Document document = loadDocument(file);
+
+        if (document == null) {
+            return null;
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: all the time you will need to be worried about a NPEx,  perhaps a throw new FileNotFoundException   to handle this alternate flows

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        NodeList nodes = document.getChildNodes();
+        Element rootElement = (Element)nodes.item(1);
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: you could write  nodes.iterator.next()  as well

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        Map<String, Role> roles = parseRoles(rootElement);
+
+        return parseGroups(rootElement, roles);
+    }
+
+    Map<String, Group> loadFromList(List<String> list) {
+        Map<String, Group> groups = new HashMap<>();
+
+        for (String groupName : list) {
+            Permission permission = new Permission();
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: you can add this to a Constructor too ,  so you write:  new Permission(Admin);  or a factory method...
Permission.forAdmin() that returns it - same goes to other models

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class GroupsLoader {
+
+    private static final String DEFAULT_ROLE_NAME = "admin";
+    private static final String XML_DTD_FILE_PATH = "/azkaban-groups.dtd";
+
+    Map<String, Group> loadFromFile(String filePath) {
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: method is exposed ( is it public ? private ? )

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+        return roles;
+    }
+
+    private Map<String, Group> parseGroups(Element rootElement, Map<String, Role> roles) {
+        NodeList groupNodes = rootElement.getElementsByTagName("group");
+        Map<String, Group> groups = new HashMap<>();
+
+        for (int i = 0; i < groupNodes.getLength(); i++) {
+            NamedNodeMap attributes = groupNodes.item(i).getAttributes();
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: why not iterate with a for-each ? on groupNodes?

File: src/main/java/net/researchgate/azkaban/LdapUserManager.java [hunk 9]
```diff
@@ -114,38 +144,80 @@ public User getUser(String username, String password) throws UserManagerExceptio
...
-        if (ldapAllowedGroups.size() == 0) {
+    private User createUserFromEntry(Entry entry) throws UserManagerException, LdapException {
+        Attribute idAttribute = entry.get(ldapUserIdProperty);
+        Attribute emailAttribute = null;
+
+        if (idAttribute == null) {
+            throw new UserManagerException("Invalid id property name " + ldapUserIdProperty);
+        }
+
+        if (ldapUEmailProperty.length() > 0) {
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/LdapUserManager.java]: maybe there is a on  ldapUEmailProperty.isEmpty()?

File: src/main/java/net/researchgate/azkaban/LdapUserManager.java [hunk 10]
```diff
@@ -114,38 +144,80 @@ public User getUser(String username, String password) throws UserManagerExceptio
...
+        }
+
+        return user;
+    }
+
+    private void addGroupsToUser(User user, Entry userEntry, LdapConnection connection) throws LdapException {
+        if (ldapEmbeddedGroups) {
+            for (String groupName: groups.keySet()) {
+
+                String groupDN = "CN=" + groupName + "," + ldapGroupSearchBase;
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/LdapUserManager.java]: strings are immutable, perhaps use a Stringbuilder.append for this operation ?

File: src/main/java/net/researchgate/azkaban/LdapUserManager.java [hunk 11]
```diff
@@ -114,38 +144,80 @@ public User getUser(String username, String password) throws UserManagerExceptio
...
+        Attribute userDn = userEntry.get(ldapUserIdProperty);
+
+        for (String groupName: groups.keySet()) {
+            Entry result = connection.lookup("CN=" + groupName + "," + ldapGroupSearchBase);
+
+            if (result == null) {
+                continue;
+            }
+
+            Attribute members = result.get("memberuid");
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/LdapUserManager.java]: memberuid  could be a static final constant string

File: src/main/java/net/researchgate/azkaban/LdapUserManager.java [hunk 12]
```diff
@@ -114,38 +144,80 @@ public User getUser(String username, String password) throws UserManagerExceptio
...
-
-		if (members == null) {
-		    return false;
-		}
-
-		return members.contains(username.toString());
-	    }
-	}
+
+        for (String groupName : user.getGroups()) {
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/LdapUserManager.java]: wasn't a validateGroup method for that ?

File: src/test/java/net/researchgate/azkaban/GroupsLoaderTest.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+        fail("GroupsLoader should throw an exception when the file is malformed");
+    }
+
+    @Test
+    public void testValidFile() {
+        String filePath = getClass().getResource("/azkaban-groups.xml").getFile();
+        GroupsLoader loader = new GroupsLoader();
+        loader.loadFromFile(filePath);
+
```
Comment (rafaelcavazin-coolblue) [inline src/test/java/net/researchgate/azkaban/GroupsLoaderTest.java]: this empty lines can be avoided

File: src/main/java/net/researchgate/azkaban/Group.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
+package net.researchgate.azkaban;
+
+import azkaban.user.Role;
```
Comment (rafaelcavazin-coolblue) [inline src/main/java/net/researchgate/azkaban/Group.java]: if your code has to be compliant with java 7 and you still would like to lambda-up some stuff .. you can add Google's Guava as dependency ... there are lots of cool stuffs to make your code look sexy ;)

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java (hunk 8 above)
Comment (hugoboos) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: Foreach is not supported on `NodeList`.

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java (hunk 7 above)
Comment (hugoboos) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: Method can be package private. It is only used inside the package.

File: src/main/java/net/researchgate/azkaban/GroupsLoader.java (hunk 5 above)
Comment (hugoboos) [inline src/main/java/net/researchgate/azkaban/GroupsLoader.java]: `NodeList` does not have a `. iterator()`.

File: src/main/java/net/researchgate/azkaban/LdapUserManager.java (hunk 12 above)
Comment (hugoboos) [inline src/main/java/net/researchgate/azkaban/LdapUserManager.java]: There was. But I rebuild it into `addGroupsToUser()`.

**Comment by bashofmann** [comment]
@hugoboos Maybe there is another public ldap server for testing somewhere out there. But at least the existing tests should not fail in order to merge this

**Pull request closed without merging**